File: Applied Course/4.Problem Solving/1.Problems on Array/13.Inversions in an array.cpp
#include<bits/stdc++.h>
using namespace std;
int mergeAndCount(vector<int>& A,int low,int mid,int high)
{
	int count=0,i=low,j=mid+1,k=low;
	vector<int> temp(A);
	while(i<=mid && j<=high)
	{
		if(A[i]>A[j])
		{
			count+=(mid-i+1);
			temp[k++]=A[j++];
		}
		else
			temp[k++]=A[i++];
	}
	while(i<=mid)
		temp[k++]=A[i++];
	while(j<=high)
		temp[k++]=A[j++];
	for(int i=low;i<=high;i++)
		A[i]=temp[i];
	return count;
}
int countInversions(vector<int>& A,int low,int high)
{
	if(low>=high)
		return 0;
	int mid=low+(high-low)/2;
	int l=countInversions(A,low,mid);
	int r=countInversions(A,mid+1,high);
	int m=mergeAndCount(A,low,mid,high);
	return l+r+m;
}
int main()
{
	vector<int> v;
	int n;
	cin>>n;
	for(int i=0;i<n;i++)
	{
		int x;
		cin>>x;
		v.push_back(x);
	}
	cout<<countInversions(v,0,v.size()-1);
	return 0; 
}

File: Applied Course/4.Problem Solving/1.Problems on Array/13.Inversions in an array.py
def merge(arr,temp,start,end):
    inv_count=0
    if start < end:
        
        mid=(start+end)//2
        
        inv_count+=merge(arr,temp,start,mid)
        
        inv_count+=merge(arr,temp,mid+1,end)
        
        inv_count+=mergeAndCount(arr,temp,start,mid,end)
        
    return inv_count

def mergeAndCount(arr,temp,start,mid,end):
    
    i=start
    j=mid+1
    k=start
    
    inv_count=0
    while i<=mid and j<=end:
        
        if(arr[i] > arr[j]):
            temp[k] = arr[j] 
            inv_count += (mid-i + 1) 
            k += 1
            j += 1
        else:
            temp[k] = arr[i] 
            k += 1
            i += 1
    
    while i<=mid:
        temp[k] = arr[i] 
        k += 1
        i += 1
    
    while j<=end:
        temp[k] = arr[j] 
        k += 1
        j += 1
    
    for i in range(start,end+1):
        arr[i]=temp[i]
        
    return inv_count

if __name__=='__main__':
    arr = [1, 2, 6, 4, 5] 
    n = len(arr) 
    temp=[None]*10
    count=merge(arr,temp,0,n-1)
    print(count)
